Bug fixes for package netcdf-cxx4 on macOS, align netcdf-cxx4 with netcdf-fortran

climbfuji opened the pull request:
This PR fixes a problem building netcdf-cxx4 (all the way up to the latest public release 4.3.1) on macOS with its semi case-sensitive file systems. An additional bug fix is required to get the correct library dependency on netcdf-c into the netcdf-cxx4 build on macOS, which uses `libtool`.

The PR also aligns the netcdf-cxx4 package config with the (better maintained) netcdf-fortran package config.

Fixes https://github.com/spack/spack/issues/29245
Fixes https://github.com/spack/spack/issues/26559

This PR was tested on macOS Monterey with llvm-clang@13.0.0 (and gfortran@11.2.0), as well as on a Cray (NOAA-GFDL Gaea) with intel@2021.3.0.

climbfuji committed:
Bug fixes for package netcdf-cxx4 so that it builds on macOS semi case-sensitive filesystems; build package netcdf-cxx4 consistently with netcdf-fortran

skosukhin commented on var/spack/repos/builtin/packages/netcdf-cxx4/package.py:
```diff
@@ -24,15 +25,15 @@ class NetcdfCxx4(AutotoolsPackage):
     variant('shared', default=True, description='Enable shared library')
     variant('pic', default=True, description='Produce position-independent code (for shared libs)')
-    variant('doxygen', default=True, description='Enable doxygen docs')
+    variant('doc', default=False, description='Enable doxygen docs')
```
I support this renaming. Thank you!

skosukhin commented on var/spack/repos/builtin/packages/netcdf-cxx4/package.py:
```diff
@@ -43,7 +44,17 @@ def flag_handler(self, name, flags):
         elif name == 'cppflags':
             flags.append('-I' + self.spec['netcdf-c'].prefix.include)
-
+        elif name == 'ldflags':
```
I think we don't need this because Spack does not install the `*.la` files for quite some time already.  Please, feel free to remove this logic from `netcdf-fortran` too.

skosukhin commented on var/spack/repos/builtin/packages/netcdf-cxx4/package.py:
```diff
@@ -43,7 +44,17 @@ def flag_handler(self, name, flags):
+            # can correctly link to all the dependencies even when the
+            # building takes place outside of Spack environment, i.e.
+            # without Spack's compiler wrappers.
+            config_flags = [self.spec['netcdf-c'].libs.search_flags]
```
The variable `config_flags` is set but seems not to be used anywhere.

climbfuji replied:
This is copied from the netcdf-fortran package as well, but I missed returning the flags. Since you said it's not needed, I'll remove it here and in netcdf-fortran. Yes, I noticed that there are no `.la` files and I was wondering why netcdf-fortran had this code.

skosukhin commented on var/spack/repos/builtin/packages/netcdf-cxx4/package.py:
```diff
@@ -43,7 +44,17 @@ def flag_handler(self, name, flags):
+            # On macOS, we also need to add this to ldflags
+            if self.spec.satisfies("platform=darwin"):
+                flags = [self.spec['netcdf-c'].libs.search_flags, 
+                         self.spec['netcdf-c'].libs.link_flags] + flags
```
I am curious to learn why you think `link_flags` should be added. As far as I know, the configure script appends `-lnetcdf` automatically.

climbfuji replied:
Definitely needed on macOS (in this or another way). I built netcdf-c with spack and netcdf-fortran as-is in the current spack repo, but netcdf-cxx4 did not have the -L.../netcdf-c/lib -lnetcdf in the linker line.

skosukhin commented on var/spack/repos/builtin/packages/netcdf-cxx4/package.py:
```diff
@@ -56,6 +67,15 @@ def libs(self):
+        if '+mpi' in netcdf_c_spec or '+parallel-netcdf' in netcdf_c_spec:
+            config_args.append('CC=%s' % self.spec['mpi'].mpicc)
+            config_args.append('FC=%s' % self.spec['mpi'].mpifc)
+            config_args.append('F77=%s' % self.spec['mpi'].mpif77)
```
I doubt that setting `FC` and `F77` is needed for this package. I would add setting `CXX` and check whether `CC` is needed. In any case, thanks for trying to make this consistent.

climbfuji replied:
Of course, good point. I'll remove Fortran and check if we still need to set CC.

skosukhin commented on var/spack/repos/builtin/packages/netcdf-cxx4/package.py:
```diff
@@ -71,9 +91,21 @@ def configure_args(self):
+            # Unidata has fixed this in their development track:
+            #   https://github.com/Unidata/netcdf-cxx4/commit/41c0233cb964a3ee1d4e5db5448cd28d617925fb
+            os.rename('VERSION', 'config.VERSION')
+        super(NetcdfCxx4, self).build(spec, prefix)
```
could you, please, check whether this can be replaced with (also for version `4.3.0`):
```python
patch('https://github.com/Unidata/netcdf-cxx4/commit/e7cc5bab02cf089dc79616456a0a951fee979fe9.patch',
      sha256='ba466b887c7ad2710e6d394046bc0742a2f4043d98bd21d5b08f33dc9d7f392d',
      when='@:4.3.1 platform=darwin')
```
I have used https://github.com/Unidata/netcdf-cxx4/commit/e7cc5bab02cf089dc79616456a0a951fee979fe9 because that's the commit in the `master` upstream.

climbfuji replied:
Ok, I got this - will update the sha hash in my next test:
```
==> Error: ChecksumError: sha256 checksum failed for /Users/heinzell/work/jedi-stack/spack-stack-new-joint/spack-stack-develop-20220215/cache/build_stage/spack-stage-u032oec1/e7cc5bab02cf089dc79616456a0a951fee979fe9.patch
    Expected ba466b887c7ad2710e6d394046bc0742a2f4043d98bd21d5b08f33dc9d7f392d but got 4ddf6db9dc0c5f754cb3d68b1dbef8c385cf499f6e5df8fbccae3749336ba84a
```

skosukhin replied:
My bad, should be `4ddf6db9dc0c5f754cb3d68b1dbef8c385cf499f6e5df8fbccae3749336ba84a` indeed.

skosukhin replied:
You might need to run `spack clean -m` after the update and before the attempt to build.

climbfuji replied:
Sure, I have learned (the hard way) that making spack forget isn't that easy ... The patch works just fine.

skosukhin commented:
I have just realized that I contradict myself regarding the MPI wrappers. See https://github.com/spack/spack/pull/17051, which also explains why we need to add `-lnetcdf` (not only on MacOs).

Not that I really mind building with the MPI wrappers (btw, the comment in this PR refers to the Fortran-related documentation, is there one for C++?) but I normally try to avoid overlinking and would like to learn why exactly we should use the wrappers.

climbfuji commented:
> I have just realized that I contradict myself regarding the MPI wrappers. See #17051, which also explains why we need to add `-lnetcdf` (not only on MacOs).
> 
> Not that I really mind building with the MPI wrappers (btw, the comment in this PR refers to the Fortran-related documentation, is there one for C++?) but I normally try to avoid overlinking and would like to learn why exactly we should use the wrappers.

I can try building without the MPI wrappers. Maybe that problem is solved by linking against the netcdf-c library. I remember I got the missing MPI symbols first, so I put the MPI wrappers in place, before I ran into and fixed the netcdf-c linker problem.

skosukhin commented:
@climbfuji thanks, please, keep in mind that the direct dependency on MPI might be unintended. For example, see https://github.com/spack/spack/pull/6632

climbfuji commented:
> @climbfuji thanks, please, keep in mind that the direct dependency on MPI might be unintended. For example, see #6632

Without any MPI wrapper, I get this:
```
     393    In file included from /Users/heinzell/work/jedi-stack/spack-stack-new-joint/spack-stack-develop-20220215/envs/jedi-fv3.netcdfcxxtest.macos/install/clang/13.0.0/hdf5-1.12.1-i23lyyj/include/hdf5.h:22:
  >> 394    /Users/heinzell/work/jedi-stack/spack-stack-new-joint/spack-stack-develop-20220215/envs/jedi-fv3.netcdfcxxtest.macos/install/clang/13.0.0/hdf5-1.12.1-i23lyyj/include/H5public.h:63:10: fatal error: 'mpi.h' file not found
     395    #include <mpi.h>
     396             ^~~~~~~
```
Will try with just the mpicxx wrapper next.

climbfuji commented:
> > @climbfuji thanks, please, keep in mind that the direct dependency on MPI might be unintended. For example, see #6632
> 
> Without any MPI wrapper, I get this:
> 
> ```
>      393    In file included from /Users/heinzell/work/jedi-stack/spack-stack-new-joint/spack-stack-develop-20220215/envs/jedi-fv3.netcdfcxxtest.macos/install/clang/13.0.0/hdf5-1.12.1-i23lyyj/include/hdf5.h:22:
>   >> 394    /Users/heinzell/work/jedi-stack/spack-stack-new-joint/spack-stack-develop-20220215/envs/jedi-fv3.netcdfcxxtest.macos/install/clang/13.0.0/hdf5-1.12.1-i23lyyj/include/H5public.h:63:10: fatal error: 'mpi.h' file not found
>      395    #include <mpi.h>
>      396             ^~~~~~~
> ```
> 
> Will try with just the mpicxx wrapper next.

Same problem with just the MPICXX wrapper, as expected.

climbfuji committed:
Update netcdf-cxx4 and netcdf-fortran based on code review

climbfuji committed:
No need to import os anymore

climbfuji committed:
Remove unused config_flags from var/spack/repos/builtin/packages/netcdf-fortran/package.py

climbfuji committed:
Trim whitespace in var/spack/repos/builtin/packages/netcdf-cxx4/package.py

skosukhin commented:
@climbfuji could you, please, tell me which MPI library you use? I find it strange that Spack compiler wrapper does not inject the required `-I/path/to/mpi_h` flag.

climbfuji commented:
> @climbfuji could you, please, tell me which MPI library you use? I find it strange that Spack compiler wrapper does not inject the required `-I/path/to/mpi_h` flag.

mpich-3.4.3 installed via homebrew (i.e. an external package for spack)

skosukhin committed:
netcdf-fortran: avoid building without the optimization flags

skosukhin committed:
netcdf-cxx4: do not enforce autoreconf

This was a rudiment from the times when the package was fetched
with git, which broke timestamp order of the automatically
generated Autoconf files.

skosukhin committed:
netcdf-cxx4: inject PIC flags for C++ when '+pic'

skosukhin committed:
netcdf-cxx4: inject C/CXXFLAGS via the wrapper

skosukhin committed:
netcdf-cxx4: fix the underlinking problem for platforms other than darwin

skosukhin committed:
netcdf-cxx4: remove redundant extention of CPPFLAGS

skosukhin committed:
netcdf-cxx4: fixup for the MPI-related problem

skosukhin committed:
netcdf-cxx4: a bit more laconic syntax

skosukhin committed:
netcdf-cxx4: add/update the comments

skosukhin committed:
netcdf-cxx4: remove variant 'static'

This makes it consistent with other packages from the NetCDF
constellation: always build the static libraries and additionally
build the shared ones when '+shared'.

skosukhin committed:
netcdf-cxx4: do not configure --with/--without-pic

This makes it consistent with other packages from the NetCDF
constellation: build the shared libraries with the PIC flag and
the static ones without it (the default for Autotools) when
'~pic', and build the static libraries with PIC when '+pic' (to
make them injectable into other shared libraries).

skosukhin committed:
netcdf-cxx4: run the tests serially

skosukhin committed:
netcdf-cxx4: build the plugins only when the tests are run

skosukhin committed:
netcdf-cxx4: update property 'libs'

skosukhin committed:
eccodes: replace the patch with the post-configure method

skosukhin committed:
eccodes: update comments

climbfuji committed:
Merge pull request #4 from skosukhin/bugfix/netcdf_cxx4_macos

Bugfix/netcdf cxx4 macos

climbfuji committed:
Merge branch 'develop' into bugfix/netcdf_cxx4_macos

climbfuji committed:
Merge branch 'develop' into bugfix/netcdf_cxx4_macos

climbfuji committed:
Merge branch 'develop' of https://github.com/spack/spack into bugfix/netcdf_cxx4_macos

climbfuji committed:
Merge branch 'bugfix/netcdf_cxx4_macos' of https://github.com/climbfuji/spack into bugfix/netcdf_cxx4_macos

climbfuji commented:
@skosukhin @WardF I think this is ready to merge with your (Sergey's updates). Would you be able to review this? Thanks a lot in advance!

climbfuji commented:
@skosukhin @WardF I have tested this to the bones on numerous systems (macOS, Linux, Cray) with several compilers (clang, gnu, intel) in the last ten days. Can we get this merged? Thanks a lot in in advance for your time.

climbfuji commented:
@skosukhin @WardF I am sorry that I have to keep asking, but can we get this merged, please? It passed all the CI tests and works on all our platforms. I just updated the branch once again ... thanks very much!

climbfuji committed:
Merge branch 'develop' into bugfix/netcdf_cxx4_macos

skosukhin commented:
@climbfuji none of us can actually do the merge. All I can do is write LGTM.
@haampie I've picked you almost randomly :) could you, please, merge this or poke someone else who can do it? Thanks!

climbfuji commented:
> @climbfuji none of us can actually do the merge. All I can do is write LGTM.
> @haampie I've picked you almost randomly :) could you, please, merge this or poke someone else who can do it? Thanks!

Thanks, @skosukhin - appreciate your time.

climbfuji commented:
Any update on this PR?

scheibelp commented on var/spack/repos/builtin/packages/netcdf-fortran/package.py:
```diff
@@ -81,15 +79,11 @@ def flag_handler(self, name, flags):
-            # could correctly link to all the dependencies even when the
-            # building takes place outside of Spack environment, i.e.
-            # without Spack's compiler wrappers.
-            config_flags = [self.spec['netcdf-c'].libs.search_flags]
```
From my naive perspective, this looks problematic: this was removed but nothing appears to have been added to take its place (I see added logic in `netcdf-cxx4`'s `flag_handler` method which appears to do something similar to this, but I assume this would need it too). Presumably you removed this because it causes an error on Mac OS, but is it still needed on Linux?

skosukhin replied:
@scheibelp as you can see in the removed comment, this was only needed to keep the libtool's `.la` file consistent. But Spack removes such files for quite some time already (see #18850). As for the linking itself, Spack's compiler wrapper adds the required `-L` flags anyway.

scheibelp commented on var/spack/repos/builtin/packages/netcdf-cxx4/package.py:
```diff
@@ -19,61 +21,122 @@ class NetcdfCxx4(AutotoolsPackage):
 
-    # Usually the configure automatically inserts the pic flags, but we can
-    # force its usage with this variant.
-    variant('static', default=True, description='Enable building static libraries')
```
(question) With this change, does `+shared` build both the shared and static libraries, or just the shared libraries?

skosukhin replied:
@scheibelp the static libraries are always built and the shared only when `+shared`

scheibelp reviewed (commented):
I had a couple questions

(I can merge this once those are resolved)

scheibelp commented:
Thanks!

climbfuji commented:
Thanks indeed!
